src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -441,7 +441,9 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
             Models.RazorSemanticTokensEdit[] edits)
         {
             var updatedTokens = previousCSharpTokens.ToList();
-            foreach (var edit in edits)
+
+            // C# returns edits that are expected to be applied in reverse order.
+            foreach (var edit in edits.Reverse())
```

Comment:
So Roslyn produces edits moving from highest offset to lowest, then reverses it to pass it to us, at which point we're reversing the edits again to apply them? It seems like we'd be taking a HARD dependency an implementation detail in Roslyn. My alternative suggestions would be:
1. Roslyn produce the edits starting from offset 0 in the first place.
2. Either:
    1. One of Razor and Roslyn are not adhering to the LSP specification in how we generate edits, and that LSP should be fixed.
    2. The VSPlatform is not adhering to the specification, and should be fixed.
    3. The specification is insufficiently specific about deltas (so that multiple implementations can adhere to the spec but be incompatible), whether the offset in a delta is from the beginning of the OLD array or the new one, if edit ordering matters, etc. They need to clarify and then we need to ensure both LSP's and VSPlatform should be updated to adhere to the updated specification.

I think the option 2 is the more correct move, but they're all speculative on learning more about where things are going wrong (and #3 is likely to take a very long time if it's needed).

Reply:
I'm pretty sure the issue here is point 3. As far as I can tell, the LSP spec doesn't specify anything about ordering or whether `start` refers to the new or old array. I believe the LSP client today interprets `start` as referring to the old array, but we probably do want to add clarification to the spec, cc @gundermanc.

As for point 1, Roslyn today does produce the edits starting from offset 0. Roslyn _used_ to provide edits the opposite way before the changes from https://github.com/dotnet/roslyn/pull/55430. The reason Razor needs to reverse the edits is to copy the behavior of the LSP client and apply Roslyn's edits in the same order they apply Roslyn's edits.

For point 2, since there's no details in the spec, I don't think the VSPlatform is doing anything wrong here since I don't think what is considered "correct" has been defined.